File: leet_code/trees/flatten_tree_into_list.js
const preorder = root => {
  if (!root) return [];
  const preorderArr = [],
        traverse = node => {
          preorderArr.push(node);
          if (node.left) traverse(node.left);
          if (node.right) traverse(node.right);
    };
  traverse(root);
  return preorderArr;
};

const flattenTreeIntoList = root => {
  if (!root) return null;
  const traversal = preorder(root);
  for (let i = 0; i < traversal.length; i++) {
    if (i === traversal.length - 1) {
      traversal[i].left = null;
      traversal[i].right = null;
    } else {
      traversal[i].left = null;
      traversal[i].right = traversal[i+1];
    }
  }
  return root;
};

File: leet_code/trees/flatten_tree_into_list.rb
def preorder(root)
  return [] if !root
  result = [root]
  result.concat(preorder(root.left))
  result.concat(preorder(root.right))
  result
end


def flatten_tree_into_list(root)
  return nil if !root
  traversal = preorder(root)
  traversal.each_with_index do |node, idx|
    if node == traversal[-1]
      node.left, node.right = nil, nil
    else
      node.left, node.right = nil, traversal[idx+1]
    end
  end
  root
end
